pkg/client/transport/config.go

```diff
@@ -0,0 +1,101 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package transport
+
+import "net/http"
+
+// Config holds various options for establishing a transport.
+type Config struct {
```

Comment:
I'm still having a hard time seeing the benefit of deeply nested config structs... presumably, we would eventually want client.Config to contain transport.Config, at which point we'd have this:

```
client.Config
  transport.Config
    transport.AuthConfig
      transport.TokenAuthConfig
        Token string
```

That seems like a lot of levels. Just like you have the `UserAgent string` field that drives the creation of the user-agent round tripper wrapper, it seems simpler to keep the username,password fields to drive the basic auth wrapper, and the token field to drive the bearer auth wrapper.

The value seems to be in the helper functions you are checking to drive the wrapper creation (like `hasBasicAuth`, etc), and those could simply check the fields directly in the transport config, they don't needs structs within structs within structs.